### Bonus 3:
Přidej Sektor (poslední sloupec) jako vysvětlující proměnnou s využitím One Hot Encoding. Podívej se, které sektory mají kladný koeficient a které sektory mají záporný koeficient.

In [1]:
import pandas as pd
import statsmodels.formula.api as smf

data = pd.read_csv("df.csv")
data

Company    CR    DA     FL     OPM      RT   TAT     DE      Y     PT  \
0    2Q3DU  1.19  0.20   2.21  0.0770    6.66  0.40   0.43   1.95   5.93   
1    46JNq  0.29  0.64  10.14  0.2037   20.04  0.36   6.52   7.26  43.48   
2    4ISOt  2.93  0.04   1.43  0.3055    6.55  0.72   0.06   7.78  18.38   
3    5AWwu  1.03  0.25   3.55  0.0708  146.43  1.80   0.88   6.17   5.15   
4    5hbZR  1.20  0.47   0.00  0.1677   30.92  0.93   0.00   0.00  17.06   
..     ...   ...   ...    ...     ...     ...   ...    ...    ...    ...   
95   yhCH3  1.50  0.37   2.33  0.2557    4.92  0.41   0.85   5.18   6.83   
96   YqdX2  0.70  0.27   2.57  0.2363   16.11  0.64   0.69   7.40   2.70   
97   yUjyz  1.49  0.12   2.36  0.2576    8.64  0.51   0.28   5.99   7.59   
98   YX515  2.30  0.64  25.03  0.3758    7.48  0.31  16.08  41.57  10.54   
99   zCRGR  1.40  0.21   2.35  0.2391    7.08  0.45   0.50   3.35   5.53   

                    Sector  
0              Industrials  
1   Communication Services  
2   Communication Services  
3       Consumer Defensive  
4        Consumer Cyclical  
..                     ...  
95              Healthcare  
96      Consumer Defensive  
97              Technology  
98              Technology  
99              Healthcare  

[100 rows x 11 columns]

One Hot Encoding poprvé - Při využití get_dummies se musí oba dataframy spojit pomocí pd.concat. Když se používají kategorické proměnné v regresi, obvykle se jedna kategorie vyřazuje (v pd.get_dummies se nastaví drop_first=True). Vyřazená kategorie slouží jako referenční (ostatní se s ní porovnávají).

In [2]:
# OHE = pd.get_dummies(data["Sector"], drop_first=True)
# data_ohe = pd.concat([data, OHE], axis=1 )
# data_ohe.head()

In [3]:
# formula = "Y ~ CR + DA + FL + OPM + RT + TAT + DE + PT + data_ohe['Communication Services'] + data_ohe['Conglomerates'] + data_ohe['Consumer Cyclical'] + data_ohe['Consumer Defensive'] + data_ohe['Energy'] + data_ohe['Financial Services'] + data_ohe['Healthcare'] + data_ohe['Industrials'] + data_ohe['Technology']"

# mod = smf.ols(formula=formula, data=data)
# res = mod.fit()
# res.summary()

One Hot Encoding podruhé - Formula umí pracovat s kategorickými proměnnými, přidá se + C().

In [4]:
formula = "Y ~ CR + DA + FL + OPM + RT + TAT + DE + PT + C(Sector)"

mod = smf.ols(formula=formula, data=data)
res = mod.fit()
res.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      Y   R-squared:                       0.714
Model:                            OLS   Adj. R-squared:                  0.655
Method:                 Least Squares   F-statistic:                     12.04
Date:                Sat, 25 Nov 2023   Prob (F-statistic):           8.25e-16
Time:                        08:45:55   Log-Likelihood:                -367.95
No. Observations:                 100   AIC:                             771.9
Df Residuals:                      82   BIC:                             818.8
Df Model:                          17                                         
Covariance Type:            nonrobust                                         
=======================================================================================================
                                          coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------------------
Intercept                              -7.4799      7.709     -0.970      0.335     -22.816       7.857
C(Sector)[T.Communication Services]    -3.7782      7.099     -0.532      0.596     -17.901      10.344
C(Sector)[T.Conglomerates]             -4.7419     12.308     -0.385      0.701     -29.226      19.742
C(Sector)[T.Consumer Cyclical]          3.0960      7.001      0.442      0.659     -10.831      17.023
C(Sector)[T.Consumer Defensive]        -0.3148      7.466     -0.042      0.966     -15.167      14.537
C(Sector)[T.Energy]                    -6.1282      7.521     -0.815      0.418     -21.089       8.833
C(Sector)[T.Financial Services]        10.5386      9.441      1.116      0.268      -8.243      29.320
C(Sector)[T.Healthcare]                -4.0329      6.689     -0.603      0.548     -17.340       9.275
C(Sector)[T.Industrials]               -7.7802      6.933     -1.122      0.265     -21.572       6.012
C(Sector)[T.Technology]                 1.2953      6.753      0.192      0.848     -12.139      14.730
CR                                      1.3840      1.266      1.093      0.277      -1.134       3.902
DA                                     -2.1773      9.928     -0.219      0.827     -21.926      17.572
FL                                      3.2184      1.120      2.873      0.005       0.990       5.447
OPM                                    17.4452      8.742      1.996      0.049       0.055      34.835
RT                                     -0.0120      0.022     -0.544      0.588      -0.056       0.032
TAT                                     7.0691      2.558      2.763      0.007       1.980      12.158
DE                                     -1.3182      2.175     -0.606      0.546      -5.645       3.009
PT                                     -0.0806      0.104     -0.776      0.440      -0.287       0.126
==============================================================================
Omnibus:                       29.150   Durbin-Watson:                   1.819
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              100.330
Skew:                           0.887   Prob(JB):                     1.64e-22
Kurtosis:                       7.575   Cond. No.                     1.16e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.16e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

V regresní analýze představují koeficienty vypočtené hodnoty, které reprezentují změnu v cílové proměnné (Y) v důsledku jednotkové změny v příslušné nezávislé proměnné (proměnných z formule).

Kladné a záporné koeficienty odrážejí směr a sílu vztahu mezi danou nezávislou proměnnou a cílovou proměnnou:

- Kladné koeficienty: Naznačují pozitivní vztah mezi nezávislou a cílovou proměnnou. To znamená, že pokud se hodnota nezávislé proměnné zvětší o jednotku, očekává se zvyšení hodnoty cílové proměnné o hodnotu koeficientu.

- Záporné koeficienty: Naznačují negativní vztah mezi nezávislou a cílovou proměnnou. To znamená, že pokud se hodnota nezávislé proměnné zvětší o jednotku, očekává se snížení hodnoty cílové proměnné o hodnotu koeficientu.

Pokud je například koeficient pro proměnnou "CR" kladný, znamená to, že s každým nárůstem hodnoty "CR" bude pravděpodobně rostoucí hodnota "Y" (cílové proměnné). Naopak, pokud je koeficient pro "DA" záporný, znamená to, že s každým nárůstem hodnoty "DA" bude pravděpodobně klesající hodnota "Y".

### Vysvětlení od kouče Vaška Čepeláka z kurzu

V tomhle případě ta regrese vypočte regresní koeficient pro každou kategorii proměnné Sector (kromě první Basic Materials, která je pak referenční). Například regresní koeficient C(Sector)[T.Communication Services] = -3.7782 znamená, že sektor Communication Services má proti sektoru Basic Materials (= referenční kategorie) hodnotu Y o 3.78 nižší. Přesněji řečeno: budu-li mít 2 stejné firmy se stejnými vlastnostmi a jedna bude ze sektoru Comunication Services a druhá z Basic Materials, bude ta první mít hodnotu Y v průměru o 3.78 nižší než ta druhá.